iroha_introspect/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+//! Module for introspecting rust types in other languages for translation.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    time::Duration,
+};
+
+/// Derive introspect. It will make your structure introspectable
+pub use iroha_introspect_derive::Introspect;
+use serde::Serialize;
+
+/// Metadata map
+pub type MetaMap = BTreeMap<String, Metadata>;
+
+/// Introspect trait
+pub trait Introspect {
+    /// Returns unique type name.
+    /// WARN: `std::any::type_name` is compiler related, so is not unique.
+    /// I guess we should change it somehow later
+    fn type_name() -> String {
+        let mut name = module_path!().to_owned();
+        name.push_str("::");
+        name.push_str(std::any::type_name::<Self>());
+        name
+    }
+
+    /// Returns info about current type. Will return map from type names to its metadata
+    fn get_introspect() -> MetaMap {
+        let mut map = MetaMap::new();
+        Self::introspect(&mut map);
+        map
+    }
+
+    /// Introspect function. Give it empty map and it will return description of types
+    /// related to this type
+    fn introspect(metamap: &mut MetaMap);
+}
+
+/// Metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub enum Metadata {
+    /// Named structure
+    NamedStruct(NamedStructMeta),
+    /// Unnamed structure
+    UnnamedStruct(UnnamedStructMeta),
+    /// Enum
+    Enum(EnumMeta),
+    /// Integer
+    Int(IntMode),
+    /// String
+    String,
+    /// Bool
+    Bool,
+    /// Vector with type
+    Vec(String),
+    /// Option with type
+    Option(String),
+    /// Map
+    Map(MapMeta),
+    /// Result
+    Result(ResultMeta),
+}
+
+/// Named structure. It has names on fields
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct NamedStructMeta {
+    /// Fields
+    pub declarations: Vec<Declaration>,
+    //todo add collection of properties meta defined in struct
+}
+
+/// Field
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct Declaration {
+    /// Field name
+    pub name: String,
+    /// Type
+    pub ty: String,
+}
+
+/// Unnamed structure. It doesn't have name of fields
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct UnnamedStructMeta {
+    /// Field types
+    pub types: Vec<String>,
+    //todo add collection of properties meta defined in struct
+}
+
+/// Enum metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct EnumMeta {
+    /// Enum variants
+    pub variants: Vec<EnumVariant>,
+}
+
+/// Enum variant
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct EnumVariant {
+    /// Enum variant name
+    pub name: String,
+    /// Its discriminant (or identifier)
+    pub discriminant: u8,
+    /// Its type
+    pub ty: Option<String>,
```

Comment:
Here you made assumption any enum variant receives zero or one parameter as tuple. It give as additional restrictions during development, one more thing to remember and check. Can we be sure such contract never will be broken in the future?

Reply:
In our style guide we don't allow more than 1 type in variant

Reply:
It is just a style guide. It can be bypassed if there will be reasons for it. Can we be sure we won't have reasons to violate our style guide?

Reply:
This requirement is stylistic and is already encoded in other macroses also.

Reply:
Ok let's pretend it something solid